src/ariel-os-rt/src/stack.rs

```diff
@@ -0,0 +1,180 @@
+//! Stack usage helpers.
+use core::{marker::PhantomData, ptr::write_volatile};
+
+use crate::arch::sp;
+
+/// Struct representing the currently active stack.
+///
+/// # Stack painting
+///
+/// [`Stack`] allows to measure the amount of stack effectively used through a technique called
+/// stack painting:
+///
+/// 1. When initializing the memory stack, it is filled with a sequence of bytes of known values:
+///    the paint.
+/// 2. This paint gets covered during execution with the values stored on stack.
+/// 3. When requested, the amount of covered paint is measured to estimate the amount of stack used
+///    during execution.
+///
+/// Note that this technique only provides a lower bound of stack usage, as the values stored in
+/// the stack may "collide" with the paint values.
+/// In the current implementation, and assuming the stack data follows a uniform distribution, this
+/// is unlikely to result in an underestimation of more than one byte.
+///
+/// # Note
+///
+/// The machinery for stack painting has a couple of assumptions:
+///
+/// 1. It is safe for an active stack to *overwrite* unused stack space from its limit (lowest
+///    address, including) to its stack pointer (not including) through raw pointers.
+/// 2. It is safe to *read* unused stack space below the raw stack pointer down to its limit (lowest address).
+/// 3. The limits of an active stack never change.
+/// 4. It is fine to specify zero for both `lowest` and `highest`, in which case usage data is invalid
+///    (always zero), but no unsafety arises.
+/// 5. `!Send + !Sync` keeps `Stack` on the stack it was created for.
```

Comment:
That's not assumption, just how the language works:
```suggestion
/// 5. `!Send` keeps `Stack` on the stack it was created for.
```